cranelift/codegen/src/isa/aarch64/inst/emit.rs

```diff
@@ -3015,6 +3015,10 @@ impl MachInstEmit for Inst {
                 // Emit the jump itself.
                 sink.put4(enc_jump26(0b000101, dest.as_offset26_or_zero()));
             }
+            &Inst::Args { .. } => {
+                // Nothing: this is a pseudoinstruction that serves
+                // only to constrain registers at a certain point.
+            }
```

Comment:
This seems semi fragile. Is there a reason that we need an instruction for this and can't just constrain the collector directly at the top of `VCodeBuilder::collect_operands` and `count_operands` (ideally factoring out a `run_collector_on_function` helper that is used by both)? This would avoid having a magic instruction that has to always be at the top of the function and remove one extra `Inst` variant we would have to match on everywhere else.

Reply:
Unfortunately, the whole allocator is built around operands living on instructions; there aren't any free-floating constraints that can exist separately. The alternative would be to inject the operands artificially as early-def operands on the first instruction in the body, but that seems worse to me.

FWIW we do the same thing in the other direction on return instructions already. (I guess that one is a little more tightly coupled because the return also actually does the control transfer.)

Another alternative could be to build a notion of "input constraints" into RA2, but the cleanest way to implement that IMHO would be to inject an artificial instruction in RA2's frontend, rather than have a separate kind of constraint, so it doesn't improve much in global complexity.

I may be missing an option here though, if you had something else in mind?